corehq/apps/app_manager/models.py

```diff
@@ -5947,7 +5951,101 @@ def save(self, *args, **kwargs):
+        """
+        expire cache for all location and its descendants for the app
+        why? : Latest enabled release for a location is dependent on restrictions added for
+        itself and its ancestors. Hence we expire the cache for location and its descendants for which the
+        latest enabled release would depend on this location
+        :param domain: domain name
+        :param location_id: location for which the restriction is being updated
+        :param app_id: app id
+        """
+        location = SQLLocation.active_objects.get(location_id=location_id)
+        location_and_descendats = location.get_descendants(include_self=True)
+        for loc in location_and_descendats:
+            get_latest_enabled_app_release.clear(domain, loc.location_id, app_id)
+
+    def clean(self):
+        enabled_release = get_latest_enabled_app_release(self.domain, self.location.location_id, self.app_id)
+        if enabled_release:
+            if enabled_release.version > self.version:
+                raise ValidationError({'version': _("Higher version {} already enabled for this application and "
+                                                    "location").format(enabled_release.version)})
+
+    @classmethod
+    def update_status(cls, domain, app_id, build_id, location_id, version, status):
+        """
+        create a new object or just set the status of an existing one with provided
+        domain, app_id, build_id, location_id and version to the status passed
+        :param domain: domain name
+        :param app_id: app id
+        :param build_id: id of the build corresponding to the version
+        :param location_id: location_id of location
+        :param version: version number
+        :param status: expected status
+        """
+        try:
+            enabled_app_release = LatestEnabledAppReleases.objects.get(
+                domain=domain, app_id=app_id, build_id=build_id, location_id=location_id, version=version
+            )
+        except cls.DoesNotExist:
+            enabled_app_release = LatestEnabledAppReleases(
+                domain=domain, app_id=app_id, build_id=build_id, location_id=location_id, version=version
+            )
+        enabled_app_release.active = status
+        enabled_app_release.full_clean()
+        enabled_app_release.activate() if status else enabled_app_release.deactivate()
+
+    def deactivate(self):
+        self.active = False
+        self.deactivated_on = datetime.datetime.utcnow()
+        self.save()
+        self.expire_cache(self.domain, self.location.location_id, self.app_id)
+
+    def activate(self):
+        self.active = True
+        self.activated_on = datetime.datetime.utcnow()
+        self.save()
+        self.expire_cache(self.domain, self.location.location_id, self.app_id)
+
+    @classmethod
+    def to_json(cls, domain, location_id, app_id, version):
+        restrictions = cls.objects.filter(domain=domain)
+        if location_id:
```

Comment:
If these arguments aren't required, I'd make them optional by using kwargs.
I think a nicer way of doing this is to use a django serializer, which you can pass a queryset: https://docs.djangoproject.com/en/2.1/topics/serialization/#serialization-formats-json
That way you can let the caller do other types of filtering without having to modify this method.

Reply:
updated to use kwargs.
I tried using both both DjangoJSONSerializer and ModelSerializer. I also had the challenge of passing `app_names` to the serializer which i could do in `ModelSerializer` but had other challenges. Going to leave this as-is but happy to spend more time if you can suggest how to make changes for this.

Reply:
sorry, that wasn't totally clear. I meant using key word arguments for the optional arguments. something like:
```py
def to_json(cls, domain, location_id=None, app_id=None, version=None):
    ...
```

However, there's something that bugs me about this method, since I would expect a `to_json` method to serialize a single instance, and I would expect to be able to serialize a whole queryset instead of having to pass in options to the method itself. 
I think I would probably have called it like this:
```py
        enabled_app_releases: [
            release.to_json() for release in
            LatestEnabledAppReleases.objects.filter(
                domain=self.domain,
                location-id=self.request.GET.get('location_id'),
                app_id=self.request.GET.get('app_id'), 
                version=self.request.GET.get('version'))
        ]
```

I think that's more explicit, is less code, and gives greater flexibility down the line.

Reply:
[commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/daea93e98f4742e83ebb49a26fc4b8eab7e4ede7)